modal/_container_io_manager.py

```diff
@@ -389,12 +389,12 @@ async def _heartbeat_handle_cancellations(self) -> bool:
             input_ids_to_cancel = response.cancel_input_event.input_ids
             if input_ids_to_cancel:
                 if self._max_concurrency > 1:
-                    current_input_ids_to_cancel = [
-                        input_id for input_id in input_ids_to_cancel if input_id in self.current_inputs
-                    ]
-                    await self._push_terminated_outputs(current_input_ids_to_cancel)
-                    for input_id in current_input_ids_to_cancel:
-                        self.current_inputs[input_id].cancel()
+                    for input_id in input_ids_to_cancel:
+                        if input_id in self.current_inputs:
+                            # Avoid issuing cancellation signals for the same inputs repeatedly.
+                            io_context = self.current_inputs[input_id]
+                            if io_context._cancel_callback:
```

Comment:
not sure about private field read here, maybe we can adjust `IOContext.cancel()` to behave idempotently?

Reply:
done